python/ql/src/experimental/semmle/python/Concepts.qll

```diff
@@ -40,6 +40,77 @@ module SystemCommandExecution {
   }
 }
 
+/**
+ * A data-flow node that performs a file system access, including reading and writing data,
+ * creating and deleting files and folders, checking and updating permissions, and so on.
+ *
+ * Extend this class to refine existing API models. If you want to model new APIs,
+ * extend `FileSystemAccess::Range` instead.
+ */
+class FileSystemAccess extends DataFlow::Node {
+  FileSystemAccess::Range range;
+
+  FileSystemAccess() { this = range }
+
+  /** Gets an argument to this file system access that is interpreted as a path. */
+  DataFlow::Node getAPathArgument() { result = range.getAPathArgument() }
+}
+
+/** Provides a class for modeling new file-system access APIs. */
+module FileSystemAccess {
+  /**
+   * A data-flow node that performs a file system access, including reading and writing data,
+   * creating and deleting files and folders, checking and updating permissions, and so on.
+   *
+   * Extend this class to model new APIs. If you want to refine existing API models,
+   * extend `FileSystemAccess` instead.
+   */
+  abstract class Range extends DataFlow::Node {
+    /** Gets an argument to this file system access that is interpreted as a path. */
+    abstract DataFlow::Node getAPathArgument();
+  }
+}
+
+/**
+ * A data-flow node that performs path normlization. This is often needed in oder
+ * to safely access paths.
+ */
+class PathNormalization extends DataFlow::Node {
+  PathNormalization::Range range;
+
+  PathNormalization() { this = range }
+}
+
+/** Provides a class for modeling new path normalization APIs. */
+module PathNormalization {
+  /**
+   * A data-flow node that performs path normlization. This is often needed in oder
+   * to safely access paths.
+   */
+  abstract class Range extends DataFlow::Node { }
+}
+
+/**
+ * A data-flow node that checks validates a path, for instance checking that it exists
+ * or that it is safe to access.
+ */
```

Comment:
I fully agree. I think I perhaps had a hierarchy in mind, inspired by [the `EscapeFunction` from Go](https://github.com/github/codeql-go/blob/2b07e6a0f4d1641ce4fc64dc2495490b2e4804b5/ql/src/semmle/go/Concepts.qll#L120-L131) where subclasses makes the string safe in different contexts.

I am not exactly sure of the benefit of such a hierarchy, though, and I guess the general class can always be added later, so we could just have the specific on for now.

In any case, the query _must_ use the concept specific to safe access.

Reply:
Perhaps the concept should be `SafeAccessCheck` (inventing the convention that barrier guards have a `Check` suffix).

Reply:
Yeah, we're in a bit of a tricky situation. We made a setup that requires us to make a general Concept for this "safe access check" (so we can model different libraries that performs this), but we're not _really_ at a point where designing a good general concept is easy, since we only really care about it in the context of that one query.

I'm not really sure how to handle this in a nice way. I'm can think of 2 approaches:

1. Make a very narrow Concept where it's obvious that this is only meant for the context of path-injection
2. Expose library modeling and only handle the safe check inside the configuration for the query.

I'm not a big fan of option 2, since that moves all the logic of "this piece of code performs a safe access check" into the query logic, which violates how we do everything else.